### PR TITLE
fix(batch): release epoch at the end of the distributed query if it contains lookup join

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

**This section will be used as the commit message. Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- As title.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/6979

### CODE REVIEW

**Review by BugenZhao — COMMENTED** [review]
This seems to be a workable fix, however, it will definitely defer the release of the pinned snapshot. 🤔 Is it possible to also record the stages with lookup join?

**Comment by chenzl25** [comment]
> This seems to be a workable fix, however, it will definitely defer the release of the pinned snapshot. 🤔 Is it possible to also record the stages with lookup join?

Even if we find out the stage of the lookup join, we can't release the pinned snapshot right after it gets scheduled. Previously we tried to release the pinned snapshot right after all scan iterators were created. I think as long as we hold the iterators, the storage can't release its corresponding epoch which is almost equivalent with releasing the epoch at the end of the query, because in most cases scan iterators end with the same time as the whole query.

**Comment by BugenZhao** [comment]
Makes sense. I've recalled it wrong since long ago, the table scan did not pin a specific version, and it's the frontend's responsibility to pin the snapshot for the whole process. 🤣

File: src/frontend/src/scheduler/distributed/query.rs [hunk 1]
```diff
@@ -247,14 +244,6 @@ impl QueryRunner {
                         stage_id
                     );
                     self.scheduled_stages_count += 1;
-                    stages_with_table_scan.remove(&stage_id);
-                    if stages_with_table_scan.is_empty() {
-                        // We can be sure here that all the Hummock iterators have been created,
```
Comment (BugenZhao) [inline src/frontend/src/scheduler/distributed/query.rs:252]: Will there be race cases previously?

**Review by BugenZhao — APPROVED** [review]
BTW, is it possible to still eagerly unpin the snapshot if there's no lookup join? Anyway, let's get this fix merged now.

**Comment by chenzl25** [comment]
> BTW, is it possible to still eagerly unpin the snapshot if there's no lookup join? Anyway, let's get this fix merged now.

It is still possible to unpin the snapshot if there's no lookup join.

File: src/frontend/src/scheduler/distributed/query.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/scheduler/distributed/query.rs:252]: It looks fine. I can't tell.

**Comment by hzxa21** [comment]
Discussed with @chenzl25 offline. Lookup join will create storage iterators during its exeuction, not during scheduling. Based on the current interface, I cannot think of a way to do early epoch unpin in this case. To do that, we may need to do either of the following:
1. Expose the `seek` interface for storage table iterator. Lookup join can create the iterator during scheduling and call `seek` during its execution. However, since we use stream for storage table iterator, `seek` is counter-intuitive.
2. Expose the concept of storage `version` and allow upper layer to create multiple iterators on top of a version. This needs a bigger refactor and adds burden on the upper layer to unpin version.

IMO, we can delay epoch unpin if the query has lookup join for now, given that we don't expect batch query to run for a long time in the serving use cases. We can optimize it when we do see long running batch queries in the future.

File: src/frontend/src/scheduler/distributed/query.rs [hunk 2]
```diff
@@ -247,14 +244,6 @@ impl QueryRunner {
                         stage_id
                     );
                     self.scheduled_stages_count += 1;
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/distributed/query.rs:246]: Suggest to add some comments for why remove the early unpin.

**Review by BowenXiao1999 — APPROVED** [review]
LGTM

**Comment by chenzl25** [comment]
I have reimplemented in this way, only delay epoch unpin for lookup join. Please take a lookup again. @BugenZhao @BowenXiao1999

**Comment by lmatz** [comment]
@BowenXiao1999 
I forget the intention of the old implementation, 
We probably wanted to make sure that compaction can still happen -> new version, but the version/snapshot held by the iterator cannot be vacuumed.
Can't remember exactly......

**Pull request merged**